src/YoctoDao/OffChain.hs

```diff
@@ -0,0 +1,195 @@
+            , scriptUpdate  = scriptUpdate cp
+            , scriptAddr = scriptAddr cp
+            , spend = spend cp
+            , spendValue = spendValue cp
+            , spendAddr = spendAddr cp
+            , minting = minting cp
+            , minted = minted cp
+            , yes = 0
+            , no = 0
+            } --}
+data SpendOwned = SpendOwned
+    { spendAddr  :: !Address
+    , spendVal   :: !Value
+    , spendDatum :: Integer
+    } deriving (Show, Generic, FromJSON, ToJSON)
+
+type YoctoSchema = Endpoint "1-CreateOwnership" Integer .\/
+                   Endpoint "2-SpendOwnedTx" SpendOwned
+        
+createOwnerId :: forall w s. ValidatorHash -> AssetClass -> AssetClass -> AssetClass -> AssetClass -> AssetClass -> Integer -> Contract w s Text ()
+createOwnerId treasury govClass nft idMaker propclass voteclass co = do
+    Contract.logInfo @String $ printf "Creating Ownership..."
+    let scriptAddr = votingPassScriptAddress treasury govClass nft idMaker propclass voteclass
+        treasuryAddr = Address.scriptHashAddress treasury
+    pkh <- Contract.ownPubKeyHash
+    now <- currentTime
+    utxos <- utxosAt scriptAddr
+    futxos <- utxosAt (Address.pubKeyHashAddress pkh)
+
+    let utxos' = Map.filter (\v -> 1 == assetClassValueOf (_ciTxOutValue v) idMaker) utxos
+        futxos' = Map.filter (\v -> 1 <= assetClassValueOf (_ciTxOutValue v) govClass) futxos
+        (nftRef,nftTx) = head $ Map.toList utxos'
+        valueToScript = assetClassValue voteclass 1 <> assetClassValue govClass co
+        ownership = Ownership
+            { owner = pkh
+            , nftSlot = now
+            , lastTransfer = now
+            }
+        oDatum = Datum $ toBuiltinData ownership
+
+        validator = votingPassValidatorScript treasury govClass nft idMaker propclass voteclass
+        validatorHash = votingPassValidatorHash treasury govClass nft idMaker propclass voteclass
+
+        lookups = Constraints.ownPubKeyHash pkh <>
+                  Constraints.mintingPolicy (policy idMaker) <>
+                  Constraints.otherScript validator <>
+                  Constraints.unspentOutputs utxos' <>
+                  Constraints.unspentOutputs futxos'
+
+        -- The line below actually causes a plutus error, this requires more investigation..
+        --tx = Constraints.mustSpendScriptOutput nftRef (Redeemer $ toBuiltinData ()) <>
+        tx = Constraints.mustMintValue (assetClassValue voteclass 1) <>
+             Constraints.mustPayToOtherScript validatorHash oDatum valueToScript
+
+    ledgerTx <- submitTxConstraintsWith @Void lookups tx
+    void $ awaitTxConfirmed $ getCardanoTxId ledgerTx
+    Contract.logInfo @String $ printf "Created Ownership: %s" (show ownership)
+
+spendOwnershipUTxO :: forall w s. ValidatorHash -> AssetClass -> AssetClass -> AssetClass -> AssetClass -> AssetClass -> SpendOwned -> Contract w s Text ()
+spendOwnershipUTxO treasury govClass nft idMaker propclass voteclass spend = do
```

Comment:
Without looking into the logic. There's a lot of code duplication from the function above. 

Would be nice to DRY them up to be able to see the differences between these two functions.

Reply:
Yeah, I think that this is definitely a great point and there's plenty of cases where DRY principles ought to apply and have not yet been applied throughout the codebase. I am trying to clean things up a bit before we go over them tonight/tomorrow morning.